nucypher/datastore/db/models.py

```diff
@@ -34,71 +37,84 @@ class DatastoreRecord(Mapping):
 
     All objects passed to the mapping must be serializble by msgpack. If the
     provided object is _not_ serializable by msgpack, then it must be given a
-    `FieldEncoder` with an `encoder` and `decoder` function that accepts the
+    `RecordField` with an `encoder` and `decoder` function that accepts the
     provided object as an argument, and returns a msgpack serializable type,
-    identified by the attribute name in the child class (as a private attr):
+    identified by the attribute name in the child class.
+
+    All field types that don't deserialize natively from msgpack must provide
+    encoding and decoding functions in the `RecordField` to decode from `bytes`
+    to the type, and encode from the type to `bytes`. See the tests for an
+    example using a `datetime` object.
 
     ```
     class Thing(DatastoreRecord):
-        _foobar = FieldEncoder(FoobarSerializer, FoobarDeserializer)
+        one_thing = RecordField(bytes)
+        foobar = RecordField(Foobar, encoder=FoobarSerializerToBytes,
+                                     decoder=FoobarDeserializerFromBytes)
 
     my_thing = Thing(one_thing=b'one_thing', foobar=my_foobar)
     ```
 
-    In the above example, `my_thing` will have `_foobar` as a class attribute
-    that has a `FieldEncoder` that can be used to serialize when calling
-    `my_thing.to_bytes()`, and can deserialize when calling
-    `Thing.from_bytes()`.
-    """
+    In the above example, `my_thing` will have two fields: `one_thing` and
+    `foobar`. When a `Thing` is instantiated, the `RecordField`s defined in the
+    class attributes will be _consumed and replaced_ by the _decoded_ type
+    specified in the `RecordField`.
+
+    If a passed object does not validate the `RecordField.field_type`, doesn't
+    have an associated `RecordField`, or if you try to set an attribute of a
+    record that doesn't have a `RecordField`, then a `TypeError` will raise.
 
-    def __init__(self, __decode=False, **kwargs):
-        self.__data = dict(**kwargs)
-        if __decode:
-            self._decode_data()
+    Accessing the attrs of a `DatastoreRecord` should always return the
+    decoded variants. To access the encoded variant, access the record like a
+    dict as `my_thing['foobar']` or `my_thing.get('foobar')`.
+    """
 
-    def __getattr__(self, attr):
-        return self.__data[attr]
+    def __setattr__(self, attr, value):
+        try:
+            field = getattr(self, attr)
+            if not type(value) == field.field_type:
+                raise TypeError(f'The {attr} attr can only be set to a {field.field_type}.')
+        except AttributeError:
+            # TODO: Is there a better way to do this? :/
+            if not attr == '_DatastoreRecord__encoded_data':
```

Comment:
:grimacing:

Reply:
Yeah, would love a better way to do this.